CRM/Sepa/Logic/Queue/Update.php

```diff
@@ -23,121 +23,182 @@
+      'domainId'  => 1,
+    ];
+
+    return $task;
+  }
 
   /**
-   * Use CRM_Queue_Runner to do the SDD group update
-   * This doesn't return, but redirects to the runner
+   * determine the count of mandates to be investigated
+   */
+  protected static function getMandateCount($creditor_id, $sdd_mode) {
+    if ($sdd_mode == 'OOFF') {
+      $horizon = (int) CRM_Sepa_Logic_Settings::getSetting('batching.OOFF.horizon', $creditor_id);
+      $date_limit = date('Y-m-d', strtotime("+$horizon days"));
+      return CRM_Core_DAO::singleValueQuery("
+        SELECT COUNT(mandate.id)
+        FROM civicrm_sdd_mandate AS mandate
+        INNER JOIN civicrm_contribution AS contribution  ON mandate.entity_id = contribution.id
+        WHERE contribution.receive_date <= DATE('$date_limit')
+          AND mandate.type = 'OOFF'
+          AND mandate.status = 'OOFF'
+          AND mandate.creditor_id = $creditor_id;");
+    } else {
+      return CRM_Core_DAO::singleValueQuery("
+        SELECT
+          COUNT(mandate.id)
+        FROM civicrm_sdd_mandate AS mandate
+        WHERE mandate.type = 'RCUR'
+          AND mandate.status = '$sdd_mode'
+          AND mandate.creditor_id = $creditor_id;");
+    }
+  }
+
+  /**
+   * Use Civi::queue to do the SDD group update
+   *
+   * @param $mode string
+   * @return void
    */
   public static function launchUpdateRunner($mode) {
-    if (!CRM_Sepa_Logic_Settings::acquireAsyncLock('sdd_async_update_lock', SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT)) {
-      CRM_Core_Session::setStatus(ts("Cannot run update, another update is in progress!", array('domain' => 'org.project60.sepa')), ts('Error'), 'error');
+    $sdd_async_update_lock = CRM_Sepa_Logic_Settings::acquireAsyncLock(
+      'sdd_async_update_lock',
+      SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT
+    );
+
+    if (!$sdd_async_update_lock) {
+      CRM_Core_Session::setStatus(
+        ts('Cannot run update, another update is in progress!', ['domain' => 'org.project60.sepa']),
+        ts('Error'),
+        'error'
+      );
+
       $redirect_url = CRM_Utils_System::url('civicrm/sepa/dashboard', 'status=active');
       CRM_Utils_System::redirect($redirect_url);
-      return; // shouldn't be necessary
+
+      // Shouldn't be necessary
+      return;
     }
-    // create a queue
-    $queue = CRM_Queue_Service::singleton()->create(array(
-      'type'  => 'Sql',
-      'name'  => 'sdd_update',
-      'reset' => TRUE,
-    ));
-
-    // first thing: close outdated groups
-    $queue->createItem(new CRM_Sepa_Logic_Queue_Update('PREPARE', $mode));
-    $queue->createItem(new CRM_Sepa_Logic_Queue_Update('CLOSE', $mode));
-
-    // then iterate through all creditors
-    $creditors = civicrm_api3('SepaCreditor', 'get', array('option.limit' => 0));
+
+    // Create a queue
+    $queue = Civi::queue('sdd_update', [
+      'error'  => 'delete',
+      'reset'  => TRUE,
+      'runner' => 'task',
+      'type'   => 'SqlParallel',
```

Comment:
I'm a bit worried we're changing some implicit contracts on the order in which these steps are executed if we put all tasks in a single `SqlParallel` queue. The documentation mentions that for `SqlParallel`, "Items must be suitable for running out of order". While they generally seem to be picked up based on `submit_time` and `weight`, depending on your worker count, you could still end up in scenarios in which `CLEANUP` or `FINISH` is executed before the last `UPDATE`.

We may need to add something a bit more complex here, where `UPDATE` tasks run in their own `SqlParallel` queue, and `PREPARE/CLOSE/CLEANUP/FINISH` get a separate `Sql` queue that runs in order.

Since `CLEANUP` and `FINISH` need to wait for the `UPDATE` tasks to finish, we need to figure out how to implement that - one way would be to enqueue them immediately, but with a higher weight so they're not picked up immediately, but only once most of the `UPDATE` tasks are done.

Once `CLEANUP/FINISH` are picked up, we'd probably still need to check if any `UPDATE` tasks are left - if yes, we can just enqueue them again (maybe with a slightly delayed `release_time`), otherwise they do whatever they're supposed to.